### ISSUE TITLE
When upgrading to 3.0.3 from 2.18.1, invalid signature is returned from coinbase

### ISSUE THREAD

**Issue by jdgriffith** [bennycode/coinbase-pro-node#477]
![Screen Shot 2021-04-26 at 6 15 02 PM](https://user-images.githubusercontent.com/872496/116162399-53cd7100-a6bb-11eb-8f7b-ab35f0cc7c60.png)

Are there any upgrade steps? I can reproduce this locally by switching the versions.

**Comment by bennycode** [comment]
Hi @jdgriffith, can you please post the code which is throwing this error?

**Comment by jdgriffith** [comment]
![Screen Shot 2021-04-27 at 7 19 34 AM](https://user-images.githubusercontent.com/872496/116239992-eeb06480-a728-11eb-9505-9dce39a8319c.png)

![Screen Shot 2021-04-27 at 7 20 25 AM](https://user-images.githubusercontent.com/872496/116240095-0d166000-a729-11eb-947b-ec2004e8726d.png)

I also noticed you very recently created code for mutually excluding `status` and `product_id`. It's very helpful to be able to use both together so curious the thought there. 

For some reason, this works:

![Screen Shot 2021-04-27 at 7 24 46 AM](https://user-images.githubusercontent.com/872496/116240616-a6de0d00-a729-11eb-9d13-a23b7f2e664e.png)

**Comment by bennycode** [comment]
With v3.0.3 `order.getOrders` cannot be used anymore with `status` and `product_id`. These properties are mutually exclusive. I suggest to fetch by `status` and to filter the response by `product_id` in your code (using [Array.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)).

**Comment by jdgriffith** [comment]
Ok I will do that. I'm not sure that explains the invalid signature though?

**Comment by jdgriffith** [comment]
Verified locally I'm still receiving invalid signature when not using the `product_id`. 

I think* this is happening anytime I pass status to the function. 

```
    const { data: orders } = await this.client.rest.order.getOrders({
      limit,
      status: [OrderStatus.ACTIVE, OrderStatus.PENDING, OrderStatus.OPEN]
    })
```

vs

```
    const { data: orders } = await this.client.rest.order.getOrders({
      limit
    })
```

The last code example fetches data without a hitch.

**Comment by bennycode** [comment]
I was testing it actually. Can you test what happens when just using `status: ['all']`?

**Comment by jdgriffith** [comment]
This also fails for me.

```
    const { data: orders } = await this.client.rest.order.getOrders({
      limit,
      status: ['all']
    })
```

```
currentUrl: 'https://api.pro.coinbase.com/orders?status=all&limit=10'
```

```
data: { message: 'invalid signature' }
```

Something is incorrect when generating the signature for these calls (and including a status array).

**Comment by jdgriffith** [comment]
```
{
  url: '/orders?status=all',
  method: 'get',
  headers: {
    common: { Accept: 'application/json, text/plain, */*' },
    delete: {},
    get: {},
    head: {},
    post: { 'Content-Type': 'application/x-www-form-urlencoded' },
    put: { 'Content-Type': 'application/x-www-form-urlencoded' },
    patch: { 'Content-Type': 'application/x-www-form-urlencoded' }
  },
  params: { limit: 10 },
  baseURL: 'https://api.pro.coinbase.com',
  transformRequest: [ [Function: transformRequest] ],
  transformResponse: [ [Function: transformResponse] ],
  timeout: 50000,
  adapter: [Function: httpAdapter],
  xsrfCookieName: 'XSRF-TOKEN',
  xsrfHeaderName: 'X-XSRF-TOKEN',
  maxContentLength: -1,
  maxBodyLength: -1,
  validateStatus: [Function: validateStatus]
}
```

Your `config.params` does not include status from the query string for stringifying when doing the signing operation. I think this is where it is falling apart.

**Comment by bennycode** [comment]
Hmm, I was guessing that `status` is not needed in the `params` object as it's part of the url `/orders?status=all`.

**Comment by jdgriffith** [comment]
Yeah - looks like this might work if you changed this code from manually adding status to the query string:

```
    if ((query as OrderListByStatus)?.status) {
      status = '?' + (query as OrderListByStatus).status!.map(s => `status=${s}`).join('&');
      delete (query as OrderListByStatus).status;
    }
    const response = await this.apiClient.get<Order[]>(`${resource}${status}`, {
      params: query,
    });
```

**Comment by jdgriffith** [comment]
Might want to use something like:

```
https://nodejs.org/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options
```

and add a `paramsSerializer` to the `getOrders` request?

https://stackoverflow.com/questions/49944387/how-to-correctly-use-axios-params-with-arrays

Up to you but seems like there are some straightforward ways to get around the serializing of array data for axios.

**Comment by bennycode** [comment]
Thanks for your immense research. Can you try to create a PR for that? I can also help by adding the required tests for it.

**Comment by jdgriffith** [comment]
Created a PR: https://github.com/bennycode/coinbase-pro-node/pull/478

**Comment by jdgriffith** [comment]
Also, on the mutually exclusive topic, I know all too well about tripping up other people with that. Just in the last month I created some mutually exclusive env vars, complete with a fatal error when both were injected and it completely tripped up our dev-ops guys 😂 . 

Definitely not against mutually exclusive control flow in programming - I'm just not understanding why you chose it here.

**Comment by bennycode** [comment]
I chose it because I wanted to reduce the complexity as I couldn't grasp why it's failing when setting both props. So I thought it's good to have a defined state where one or the other is set.

Thanks to your support, it's now possible to set both params at the same time. Fix just got released with v3.0.4.

**Comment by jdgriffith** [comment]
oh that's awesome! Thank you sir!
